### ISSUE TITLE
Use dithering

### ISSUE THREAD

**Issue by Ser-Gen** [photopea/UPNG.js#21]
It's important for colorful images.

Source image [174285 B]
![firefox-512](https://user-images.githubusercontent.com/1856145/37969098-a1f16cca-31d8-11e8-920a-479679cd83ab.png)

After upng, web interface [69927 B]
![firefox-512 upng](https://user-images.githubusercontent.com/1856145/37969097-a1d17820-31d8-11e8-807e-71410a1123e3.png)

Result of [pngquant](https://pngquant.org/vsphotoshop.html) looks better. [69932 B]
![firefox-512 pngquant](https://user-images.githubusercontent.com/1856145/37969099-a2123d56-31d8-11e8-9b28-26ab643cad82.png)

**Comment by photopea** [comment]
I implemented Floyd-Steinberg dithering in the past, but it was not worth it.

I think we need some compression-friendly dithering. Do you know anybody who could help us?

**Comment by Ser-Gen** [comment]
pngquant uses Floyd-Steinberg modified for better color handling.

I believe, that dithering will always increase filesize because of its' random nature.
Only purpose of this feature — to pleasure our eyes.
Dithering can be hidden under flag, just like in Ps. Users will decide.

> I think we need some compression-friendly dithering. Do you know anybody who could help us?

I think, we may ask @kornelski.

**Comment by photopea** [comment]
I mean, I made three versions of image:

- A: 50 colors: 15 kB
- B: 50 colors + Dithering: 23 kB
- C: 100 colors: 22 kB

B looked as nice as C, but was slightly larger, so I thought that allowing more colors is better than dithering (both increase the file size).

I think we need dithering, that consists of some repetitive patterns, i.e. it should be "friendly" to Deflate algorithm - make B have only 20 kB (so it is still as nice as C, but smaller).

**Comment by Ser-Gen** [comment]
Oh, I see.
I don't know dithering algorithm, that can handle this case.

pngquant computes mse error, have min and max quality settings and don't write file if its' size too big or quality degrades dramatically.

Maybe you find this thread useful
https://encode.ru/threads/1757-Lossy-DEFLATE-lossy-PNG

And this project particually
https://github.com/foobaz/lossypng

**Comment by kornelski** [comment]
Yes, pngquant calculates mean square error, and applies dithering only in areas with high error. This way areas that don't need dithering don't get the extra noise.

pngquant also does edge detection (similar to Prewitt algorithm) and disables dithering on the edges. This prevents anti-aliasing look like fur.

In pngquant 90% of time is spent on K-means. If you use `--speed 10` the whole recompression (on i7 2.3Ghz) takes ~80ms dithered, 50ms undithered.
